src/commands/git-metadata/gitdb.ts

```diff
@@ -0,0 +1,308 @@
+  if (isShallow) {
+    log.info('[unshallow] Git repository is a shallow clone., unshallowing it...')
+    log.info('[unshallow] Setting remote.origin.partialclonefilter to "blob:none" to avoid fetching file content')
+    await git.addConfig('remote.origin.partialclonefilter', 'blob:none')
+    log.info(`[unshallow] Running git fetch --shallow-since="${MAX_HISTORY.oldestCommits}" --update-shallow --refetch`)
+    await git.fetch([`--shallow-since="${MAX_HISTORY.oldestCommits}"`, '--update-shallow', '--refetch'])
+    log.info('[unshallow] Fetch completed.')
+  }
+}
+
+// getKnownCommits asks the backend which of the given commits are already known
+const getKnownCommits = async (log: Logger, request: RequestBuilder, repoURL: string, latestCommits: string[]) => {
+  interface SearchCommitResponse {
+    data: Commit[]
+  }
+
+  interface Commit {
+    type: string
+    id: string
+  }
+
+  const localCommitData = JSON.stringify({
+    meta: {
+      repository_url: repoURL,
+    },
+    data: latestCommits.map((commit) => ({
+      id: commit,
+      type: 'commit',
+    })),
+  })
+  const response = await runRequest(log, 'search_commits', () =>
+    request({
+      url: '/api/v2/git/repository/search_commits',
+      headers: {
+        'Content-Type': 'application/json',
+      },
+      data: localCommitData,
+      method: 'POST',
+      timeout: API_TIMEOUT,
+    })
+  )
+  const commits = response.data as SearchCommitResponse
+  if (!commits || commits.data === undefined) {
+    throw new Error(`Invalid API response: ${response}`)
+  }
+
+  return commits.data.map((c) => {
+    if (c.type !== 'commit' || c.id === undefined) {
+      throw new Error('Invalid commit type response')
+    }
+
+    return validateCommit(c.id)
+  })
+}
+
+const validateCommit = (sha: string) => {
+  const isValidSha1 = (s: string) => /^[0-9a-f]{40}$/.test(s)
+  const isValidSha256 = (s: string) => /^[0-9a-f]{64}$/.test(s)
+
+  if (!isValidSha1(sha) && !isValidSha256(sha)) {
+    throw new Error(`Invalid commit format: ${sha}`)
+  }
+
+  return sha
+}
+
+const getObjectsToUpload = async (git: simpleGit.SimpleGit, commitsToExclude: string[]) => {
+  const rawResponse = await git.raw(
+    [
+      'rev-list',
+      '--objects',
+      '--no-object-names',
+      '--filter=blob:none',
+      `--since="${MAX_HISTORY.oldestCommits}"`,
+      'HEAD',
+    ].concat(commitsToExclude.map((sha) => '^' + sha))
+  )
+  const commitsToInclude = rawResponse.split('\n').filter((c) => c !== '')
+
+  return commitsToInclude
+}
+
+const generatePackFilesForCommits = async (log: Logger, commits: string[]): Promise<[string[], string | undefined]> => {
+  if (commits.length === 0) {
+    return [[], undefined]
+  }
+
+  const generatePackfiles = async (baseTmpPath: string): Promise<[string[], string | undefined]> => {
+    const randomPrefix = String(Math.floor(Math.random() * 10000))
+    const tmpPath = await mkdtemp(path.join(baseTmpPath, 'dd-packfiles-'))
```

Comment:
Ok, but please note that mkdtemp already generate random prefix, so we don't need do it manually

Reply:
mkdtemp generates a suffix in the directory name, while the randomPrefix here is used in each file name. At this point I'm not sure the random prefix in each individual filename is really necessary, but I'm blindly following the spec implemented in all tracers here to avoid too much rework, as it doesn't seem to hurt and we can easily revisit later.